### PR TITLE
docs(guide): add dart version of hierarchical dependency injection

### PR DESCRIPTION
**Pull request by andresaraujo**
Adds Dart version of hierarchical dependency injection guide.

@kwalrath Could you take a look?

Thanks.

### CODE REVIEW

File: public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_editor_component.dart [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      <button (click)="onSaved()">save</button>
+      <button (click)="onCanceled()">cancel</button>
+    </div>
+  </div>
+  ''')
+class HeroEditorComponent {
+  @Output() EventEmitter canceled = new EventEmitter();
+  @Output() EventEmitter saved = new EventEmitter();
+
+  RestoreService<Hero> _restoreService;
```
Comment (kasperpeulen) [inline public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_editor_component.dart]: this one can be final

File: public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+version: 0.0.1
+description: An absolute bare-bones web app.
+
+environment:
+  sdk: '>=1.0.0 <2.0.0'
+
+dependencies:
+  angular2: '2.0.0-beta.1'
+  dart_to_js_script_rewriter: '^0.1.0'
+
```
Comment (kasperpeulen) [inline public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml]: browser?

File: public/docs/_examples/hierarchical-dependency-injection/dart/web/index.html [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<!DOCTYPE html>
+
+<html>
+<head>
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Hierarchical Injector</title>
+    <script async src="main.dart" type="application/dart"></script>
+    <script async src="packages/browser/dart.js"></script>
```
Comment (kasperpeulen) [inline public/docs/_examples/hierarchical-dependency-injection/dart/web/index.html]: I think defer is preferred over async.

File: public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_editor_component.dart (hunk 1 above)
Comment (andresaraujo) [inline public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_editor_component.dart]: Also private. Good catch +1

File: public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+transformers:
+- angular2:
+    platform_directives:
+    - 'package:angular2/common.dart#CORE_DIRECTIVES'
+    - 'package:angular2/common.dart#FORM_DIRECTIVES'
```
Comment (zoechi) [inline public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml]: wouldn't juts `#COMMON_DIRECTIVES` cover both?

Comment (kwalrath) [inline public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml]: @yjbanov what would you recommend?

Comment (andresaraujo) [inline public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml]: `CORE_DIRECTIVES`

Comment (yjbanov) [inline public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml]: I'd recommend including the minimum necessary. It does not look like you need the form directives for this example app.

Comment (andresaraujo) [inline public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml]: `hero_editor_component.dart` needs an NgModel, if I don't include FORM_DIRECTIVES It won't work.

Btw If I only include `directives: const [NgModel]` in that component it also does't work, only works with FORM_DIRECTIVES. Is this a bug?

File: public/docs/_examples/hierarchical-dependency-injection/dart/.gitignore [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Files and directories created by pub
```
Comment (kwalrath) [inline public/docs/_examples/hierarchical-dependency-injection/dart/.gitignore]: You can delete this file, I think. (angular.io has its own .gitignore.)

File: public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_card_component.dart [hunk 6]
```diff
@@ -0,0 +1,16 @@
+// #docregion
+import 'package:angular2/angular2.dart';
+import 'hero.dart';
```
Comment (kwalrath) [inline public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_card_component.dart]: We should use `package:` imports for libs within the same package. (Now that we're recommending not using the `library` directive, I believe it's important not to use relative imports.)

File: public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_editor_component.dart [hunk 7]
```diff
@@ -0,0 +1,47 @@
+// #docregion
+import 'package:angular2/angular2.dart';
+import 'restore_service.dart';
```
Comment (kwalrath) [inline public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_editor_component.dart]: Ditto about using `package:` imports. I'll stop remarking on this, but it applies globally.

File: public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml [hunk 8]
```diff
@@ -0,0 +1,20 @@
+# #docregion
+name: 'hierarchical_di'
+version: 0.0.1
+description: An absolute bare-bones web app.
+
+environment:
+  sdk: '>=1.0.0 <2.0.0'
+
+dependencies:
+  angular2: '2.0.0-beta.1'
```
Comment (kwalrath) [inline public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml]: I didn't realize we were on beta.1!

**Comment by andresaraujo** [comment]
I think made all the corrections, I end up using: (Thanks to zoechi)

```
    platform_directives:
    - 'package:angular2/common.dart#COMMON_DIRECTIVES'
```

Do you want me to change this for a directive FORM_DIRECTIVES in the `hero_editor_component.dart`?

@kwalrath

File: public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_card_component.dart (hunk 6 above)
Comment (zoechi) [inline public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_card_component.dart]: @kwalrath why do you think relative imports are bad? I haven't seen any comment that they should be avoided, also not when library statements are omitted.
Would't that make Dart look more cumbersome?

Comment (kwalrath) [inline public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_card_component.dart]: There's a risk of importing the same library multiple times, iirc.

File: public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml (hunk 4 above)
Comment (yjbanov) [inline public/docs/_examples/hierarchical-dependency-injection/dart/pubspec.yaml]: Either a bug or possible you forgot an import?

**Comment by kwalrath** [comment]
@andresaraujo I think it's better to keep everything in platform_directives.

I don't know whether it's better to use COMMON_DIRECTIVES or CORE_DIRECTIVES & FORM_DIRECTIVES. @yjbanov?

**Comment by kwalrath** [comment]
This looks good to me, as far as I can tell. @sanfordredlich can you do a technical review?

**Comment by kwalrath** [comment]
@andresaraujo: You don't need to squash commits when you upload. I'll take care of that when I merge, and it'd be nice to be able to more easily see the diffs between versions.

**Comment by sanfordredlich** [comment]
LGTM. Trivial: you might want to match case on the power:

+ new Hero()+ ..name = "RubberMan"+ ..power = 'flexibility',+ new Hero()+
..name = "Tornado"+ ..power = 'Weather changer'

**Comment by yjbanov** [comment]
@kwalrath 

> I don't know whether it's better to use COMMON_DIRECTIVES or CORE_DIRECTIVES & FORM_DIRECTIVES. @yjbanov?

I have no preference

**Comment by kwalrath** [comment]
LGTM! I'll merge this.

**Pull request closed without merging**